File: src/test/java/nl/hand/made/vending/machine/controller/VendingControllerTest.java
package nl.hand.made.vending.machine.controller;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.boot.web.server.LocalServerPort;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.http.client.HttpComponentsClientHttpRequestFactory;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
class VendingControllerTest {

    @LocalServerPort
    private int port;

    @Autowired
    private TestRestTemplate restTemplate;

    @BeforeEach
    void setup() {
        restTemplate.getRestTemplate().setRequestFactory(new HttpComponentsClientHttpRequestFactory());
    }

    @Test
    void depositBuyerAccessOnly() {
        // Try depositing for a BUYER role
        ResponseEntity<Void> response = this.restTemplate.withBasicAuth("pinkunicorn", "unicornmeat")
                .postForEntity("http://localhost:" + port + "/deposit", 20, Void.class);
        Assertions.assertEquals(HttpStatus.OK, response.getStatusCode());

        // Won't work for ADMIN role
        ResponseEntity<Void> unauthorizedResponse = this.restTemplate.withBasicAuth("admin", "admin")
                .postForEntity("http://localhost:" + port + "/deposit", 5, Void.class);
        Assertions.assertEquals(HttpStatus.FORBIDDEN, unauthorizedResponse.getStatusCode());
    }

    @Test
    void buyProduct() {
        // Try buying with no deposit
        ResponseEntity<String> response = this.restTemplate.withBasicAuth("bluebear", "dancingbear")
                .postForEntity("http://localhost:" + port + "/buy", 1, String.class);
        Assertions.assertEquals(HttpStatus.BAD_REQUEST, response.getStatusCode());

        // Try buying with deposit, correct change
        ResponseEntity<Integer[]> succesResponse = this.restTemplate.withBasicAuth("orangeorange", "fruitfly")
                .postForEntity("http://localhost:" + port + "/buy", 1, Integer[].class);
        Assertions.assertEquals(HttpStatus.OK, succesResponse.getStatusCode());
        Assertions.assertNotNull(succesResponse.getBody());
        Assertions.assertEquals(1, succesResponse.getBody().length);
        Assertions.assertEquals(5, succesResponse.getBody()[0]);
    }

    @Test
    void resetDeposit() {
        // Try resetting with no deposit
        ResponseEntity<Integer[]> noChangeResponse = this.restTemplate.withBasicAuth("bluebear", "dancingbear")
                .postForEntity("http://localhost:" + port + "/reset", null, Integer[].class);
        Assertions.assertEquals(HttpStatus.OK, noChangeResponse.getStatusCode());
        Assertions.assertNotNull(noChangeResponse.getBody());
        Assertions.assertEquals(0, noChangeResponse.getBody().length);

        // Try resetting with some deposit
        ResponseEntity<Integer[]> changeResponse = this.restTemplate.withBasicAuth("rosie", "purple")
                .postForEntity("http://localhost:" + port + "/reset", null, Integer[].class);
        Assertions.assertEquals(HttpStatus.OK, changeResponse.getStatusCode());
        Assertions.assertNotNull(changeResponse.getBody());
        Assertions.assertEquals(1, changeResponse.getBody().length);
        Assertions.assertEquals(20, changeResponse.getBody()[0]);
    }
}

File: src/test/kotlin/nl/hand/made/vending/machine/controller/VendingControllerTest.kt
package nl.hand.made.vending.machine.controller

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.boot.web.server.LocalServerPort
import org.springframework.http.HttpStatus
import org.springframework.http.client.HttpComponentsClientHttpRequestFactory

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
class VendingControllerTest @Autowired constructor(
    private val testRestTemplate: TestRestTemplate
) {

    @LocalServerPort
    private val port: Int = 0

    @BeforeEach
    fun setup() {
        testRestTemplate.restTemplate.requestFactory = HttpComponentsClientHttpRequestFactory()
    }

    @Test
    fun depositBuyerAccessOnly() {
        // Try depositing for a BUYER role
        var response = testRestTemplate.withBasicAuth("pinkunicorn", "unicornmeat")
            .postForEntity("http://localhost:$port/deposit", 20, Void::class.java)
        assertEquals(HttpStatus.OK, response.statusCode)

        // Won't work for ADMIN role
        response = testRestTemplate.withBasicAuth("admin", "admin")
            .postForEntity("http://localhost:$port/deposit", 5, Void::class.java)
        assertEquals(HttpStatus.FORBIDDEN, response.statusCode)
    }

    @Test
    fun buyProduct() {
        // Try buying with no deposit
        val response = testRestTemplate.withBasicAuth("bluebear", "dancingbear")
            .postForEntity("http://localhost:$port/buy", 1, String::class.java)
        assertEquals(HttpStatus.BAD_REQUEST, response.statusCode)

        // Try buying with deposit, correct change
        val successResponse = testRestTemplate.withBasicAuth("orangeorange", "fruitfly")
            .postForEntity("http://localhost:$port/buy", 1, Array<Int>::class.java)
        assertEquals(HttpStatus.OK, successResponse.statusCode)
        assertNotNull(successResponse.body)
        assertEquals(1, successResponse.body!!.size)
        assertEquals(5, successResponse.body!![0])
    }

    @Test
    fun resetDeposit() {
        // Try resetting with no deposit
        var response = testRestTemplate.withBasicAuth("bluebear", "dancingbear")
            .postForEntity("http://localhost:$port/reset", null, Array<Int>::class.java)
        assertEquals(HttpStatus.OK, response.statusCode)
        assertNotNull(response.body)
        assertEquals(0, response.body!!.size)

        // Try resetting with some deposit
        response = testRestTemplate.withBasicAuth("rosie", "purple")
            .postForEntity("http://localhost:$port/reset", null, Array<Int>::class.java)
        assertEquals(HttpStatus.OK, response.statusCode)
        assertNotNull(response.body)
        assertEquals(1, response.body!!.size)
        assertEquals(20, response.body!![0])
    }
}